### PR TITLE
Update 03_dcsa_im_v3_0.sql

### PR DESCRIPTION
**Pull request by HenrikHL**
Expected Arrival Date "range" added in accordance with the IFS

### CODE REVIEW

**Review by douradofrazer — APPROVED** [review]

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 1]
```diff
@@ -264,6 +264,8 @@ CREATE TABLE dcsa_im_v3_0.booking (
     incoterms varchar(3) NULL REFERENCES dcsa_im_v3_0.incoterms(incoterms_code),
     invoice_payable_at varchar(100) NULL REFERENCES dcsa_im_v3_0.location(id),
     expected_departure_date timestamp NULL,
+    expected_arrival_date_start timestamp NULL,
+    expected_arrival_date_end timestamp NULL,
```
Comment (nt-as) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:268]: Should we have a `CHECK ((expected_arrival_date_start IS NULL AND expected_arrival_date_end IS NULL) OR expected_arrival_date_start < expected_arrival_date_end)` to ensure that start is less than end date when both are provided?

Comment (nt-as) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:268]: Shouldn't these be a `date` rather than a `timestamp`?

Comment (HenrikHL) [reply datamodel/initdb.d/03_dcsa_im_v3_0.sql]: Good point - added

Comment (HenrikHL) [reply datamodel/initdb.d/03_dcsa_im_v3_0.sql]: Yes - good catch. They are date in IM so changed here as well (also changed `expected_departure_date` from timestamp --> date)

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 2]
```diff
@@ -263,7 +263,9 @@ CREATE TABLE dcsa_im_v3_0.booking (
     contract_quotation_reference varchar(35) NULL,
     incoterms varchar(3) NULL REFERENCES dcsa_im_v3_0.incoterms(incoterms_code),
     invoice_payable_at varchar(100) NULL REFERENCES dcsa_im_v3_0.location(id),
-    expected_departure_date timestamp NULL,
+    expected_departure_date date NULL,
+    expected_arrival_date_start date NULL CHECK ((expected_arrival_date_start IS NULL AND expected_arrival_date_end IS NULL) OR expected_arrival_date_start <= expected_arrival_date_end),
+    expected_arrival_date_end date NULL CHECK ((expected_arrival_date_start IS NULL AND expected_arrival_date_end IS NULL) OR expected_arrival_date_start <= expected_arrival_date_end),
```
Comment (nt-as) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:268]: The CHECK is only necessary on one of them.  But I noticed I might have had a bug in my example.  This one says you need BOTH dates or NONE of them.  You cannot provide just the start or the end date.

If you are supposed to provide just one of them, then we need to make the CHECK a bit less strict.

Comment (HenrikHL) [reply datamodel/initdb.d/03_dcsa_im_v3_0.sql:268]: I am unsure about this requirement - will have to ask P7 (if endDate can be open or startDate be open...)

Comment (HenrikHL) [reply datamodel/initdb.d/03_dcsa_im_v3_0.sql]: I removed one of the checks

**Comment by HenrikHL** [comment]
Once I get the reply from P7 I will either amend this PR or merge...

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 3]
```diff
@@ -263,7 +263,9 @@ CREATE TABLE dcsa_im_v3_0.booking (
     contract_quotation_reference varchar(35) NULL,
     incoterms varchar(3) NULL REFERENCES dcsa_im_v3_0.incoterms(incoterms_code),
     invoice_payable_at varchar(100) NULL REFERENCES dcsa_im_v3_0.location(id),
-    expected_departure_date timestamp NULL,
+    expected_departure_date date NULL,
+    expected_arrival_date_start date NULL CHECK ((expected_arrival_date_start IS NULL AND expected_arrival_date_end IS NULL) OR expected_arrival_date_start <= expected_arrival_date_end),
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:267]:
```suggestion
    expected_arrival_date_start date NULL CHECK ((expected_arrival_date_start IS NULL AND expected_arrival_date_end IS NULL) OR (expected_arrival_date_start IS NOT NULL) OR (expected_arrival_date_end IS NOT NULL) OR expected_arrival_date_start <= expected_arrival_date_end),
```

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql (hunk 2 above)
Comment (HenrikHL) [reply datamodel/initdb.d/03_dcsa_im_v3_0.sql]: @nt-as reply from P7:

> In my view it should be allowed to provide one date…. i.e. the range should be optional. However at booking request stage the expected arrival date is conditional on expected departure date not being provided.

I have modified the check - please confirm it is correct ;-)

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 4]
```diff
@@ -263,7 +263,9 @@ CREATE TABLE dcsa_im_v3_0.booking (
     contract_quotation_reference varchar(35) NULL,
     incoterms varchar(3) NULL REFERENCES dcsa_im_v3_0.incoterms(incoterms_code),
     invoice_payable_at varchar(100) NULL REFERENCES dcsa_im_v3_0.location(id),
-    expected_departure_date timestamp NULL,
+    expected_departure_date date NULL,
+    expected_arrival_date_start date NULL CHECK ((expected_arrival_date_start IS NULL AND expected_arrival_date_end IS NULL) OR (expected_arrival_date_start IS NOT NULL) OR (expected_arrival_date_end IS NOT NULL) OR expected_arrival_date_start <= expected_arrival_date_end),
```
Comment (FinnAtAsseco) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:267]: I think this check will always return true and the start <= end will never be validated

Comment (FinnAtAsseco) [reply datamodel/initdb.d/03_dcsa_im_v3_0.sql:267]: The check should probably be
(expected_arrival_date_start IS NULL) OR (expected_arrival_date_end IS NULL) OR expected_arrival_date_start <= expected_arrival_date_end)

Comment (HenrikHL) [reply datamodel/initdb.d/03_dcsa_im_v3_0.sql:267]: Are you sure - what happens when both start + end are Null - won't that give an exception when doing the `expected_arrival_date_start <= expected_arrival_date_end`?

Comment (HenrikHL) [reply datamodel/initdb.d/03_dcsa_im_v3_0.sql:267]: You are right - I will change the above

Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:267]:
```suggestion
    expected_arrival_date_start date NULL CHECK ((expected_arrival_date_start IS NULL) OR (expected_arrival_date_end IS NULL) OR expected_arrival_date_start <= expected_arrival_date_end),
```

**Pull request merged**